controllers/metallb_controller.go

```diff
@@ -160,91 +161,62 @@ func (r *MetalLBReconciler) SetupWithManager(mgr ctrl.Manager, bgpType string) e
 func (r *MetalLBReconciler) syncMetalLBResources(config *metallbv1beta1.MetalLB) error {
 	logger := r.Log.WithName("syncMetalLBResources")
 	logger.Info("Start")
-	data := render.MakeRenderData()
-
-	data.Data["SpeakerImage"] = os.Getenv("SPEAKER_IMAGE")
-	data.Data["ControllerImage"] = os.Getenv("CONTROLLER_IMAGE")
-	data.Data["FRRImage"] = os.Getenv("FRR_IMAGE")
-	data.Data["IsOpenShift"] = r.PlatformInfo.IsOpenShift()
-	data.Data["NameSpace"] = r.Namespace
-	data.Data["KubeRbacProxy"] = os.Getenv("KUBE_RBAC_PROXY_IMAGE")
-	data.Data["DeployKubeRbacProxies"] = os.Getenv("DEPLOY_KUBE_RBAC_PROXIES") == "true"
 
-	data.Data["LogLevel"] = metallbv1beta1.LogLevelInfo
-	if config.Spec.LogLevel != "" {
-		data.Data["LogLevel"] = config.Spec.LogLevel
-	}
-
-	var err error
-	data.Data["MetricsPort"], err = valueWithDefault("METRICS_PORT", 7472)
+	actionConfig := new(action.Configuration)
+	chartValueOpts := &values.Options{}
+	chartValues, err := chartValueOpts.MergeValues(getter.All(r.HelmEnvSettings))
 	if err != nil {
 		return err
 	}
-	data.Data["MetricsPortHttps"], err = valueWithDefault("HTTPS_METRICS_PORT", 27472)
+	err = patchPrometheusValues(chartValues, r.Client)
 	if err != nil {
 		return err
 	}
-	data.Data["FRRMetricsPort"], err = valueWithDefault("FRR_METRICS_PORT", 7473)
+	patchControllerValues(chartValues, config, r.PlatformInfo.IsOpenShift())
+	err = patchSpeakerValues(chartValues, config)
 	if err != nil {
 		return err
 	}
-	data.Data["FRRMetricsPortHttps"], err = valueWithDefault("FRR_HTTPS_METRICS_PORT", 27473)
+	// TODO: patch rbac proxy settings if needed.
+	client := action.NewInstall(actionConfig)
+	client.ReleaseName = helmReleaseName
+	client.Namespace = r.HelmEnvSettings.Namespace()
+	client.DryRun = true
+	client.Replace = true
+	client.ClientOnly = true
+	chartPath, err := client.ChartPathOptions.LocateChart(ManifestPath, r.HelmEnvSettings)
 	if err != nil {
 		return err
 	}
-	data.Data["MLBindPort"], err = valueWithDefault("MEMBER_LIST_BIND_PORT", 7946)
+	chart, err := loader.Load(chartPath)
```

Comment:
Same for the chart. The only thing that can dinamically change is are the values.

Reply:
Done. Good suggestion about moving helm logic into new package and initializing the client and chart only once !